commerce_canadapost.module

```diff
@@ -144,27 +140,145 @@ function commerce_canadapost_commerce_shipment_form_submit($form, FormStateInter
+  if ($tracking_summary['expected-delivery-date'] != '') {
+    $values['canadapost_expected_delivery'][0]['value'] = new DrupalDateTime($tracking_summary['expected-delivery-date']);
   }
 
-  if (!empty($tracking_summary['expected-delivery-date'])) {
-    $values['field_expected_delivery'][0]['value'] = new DrupalDateTime($tracking_summary['expected-delivery-date']);
+  if ($tracking_summary['mailed-on-date'] != '') {
+    $values['canadapost_mailed_on'][0]['value'] = new DrupalDateTime($tracking_summary['mailed-on-date']);
   }
 
-  if (!empty($tracking_summary['mailed-on-date'])) {
-    $values['field_mailed_on'][0]['value'] = new DrupalDateTime($tracking_summary['mailed-on-date']);
+  if ($tracking_summary['event-location'] != '') {
+    $values['canadapost_current_location'][0]['value'] = $tracking_summary['event-location'];
   }
 
   $form_state->setValues($values);
   $shipment->set('tracking_code', $submitted_tracking_pin);
 }
 
+/**
+ * Implements hook_cron().
+ */
+function commerce_canadapost_cron() {
+  // Update tracking data for all incomplete canadapost shipments with a
+  // tracking code.
+  _commerce_canadapost_update_tracking_data();
+}
+
+/**
+ * Update tracking data for all incomplete canadapost shipments.
+ *
+ * @param array $order_ids
+ *   An array of order IDs to update the tracking data for. Leave empty to
+ *   update all orders with incomplete shipments.
+ *
+ * @return array
+ *   An array of order IDs for which the shipments were updated for.
+ */
+function _commerce_canadapost_update_tracking_data(array $order_ids = NULL) {
+  $updated_order_ids = [];
+
+  /** @var \Drupal\commerce_canadapost\Api\TrackingService $tracking_service */
+  $tracking_service = \Drupal::service('commerce_canadapost.tracking_api');
+
+  // Fetch shipments for tracking.
+  $shipments = _commerce_canadapost_fetch_shipments_for_tracking($order_ids);
+
+  foreach ($shipments as $shipment) {
+    /** @var \Drupal\commerce_shipping\Entity\ShipmentInterface $shipment */
+    // Fetch tracking summary.
+    $tracking_summary = $tracking_service->fetchTrackingSummary($shipment->getTrackingCode());
+
+    // Update the shipment fields with the tracking data.
+    $updated_order_ids[] = commerce_canadapost_update_tracking_fields($shipment, $tracking_summary);
+  }
+
+  return $updated_order_ids;
+}
+
+/**
+ * Fetch all incomplete canadapost shipments that have a tracking pin.
+ *
+ * @param array $order_ids
+ *   Only fetch shipments of specific order IDs.
+ *
+ * @return array
+ *   An array of shipment entities.
+ */
+function _commerce_canadapost_fetch_shipments_for_tracking(array $order_ids = NULL) {
+  $entity_type_manager = \Drupal::entityTypeManager();
+
+  // Query the db for the incomplete shipments.
+  $shipment_query = $entity_type_manager
+    ->getStorage('commerce_shipment')
+    ->getQuery();
+  $shipment_query
+    ->condition('type', 'canadapost')
+    ->condition('state', 'completed', '!=')
```

Comment:
im probably missing something here. but don't we want orders/shipments that have been completed?

Reply:
As I understand it, a shipment would go into complete ONLY after the person receives the delivery correct? So, until that happens, we continue to fetch and update the tracking details so that the customer knows where the package is at that moment.

Depending on how a site sets up their workflow, once the package is received, the shipment would be in completed state and we don't include them in the ^^ above query anymore.